app/utils/service_client/middleware/middleware_base.rb

```diff
@@ -0,0 +1,36 @@
+module ServiceClient
+  module Middleware
+
+    # Base class for middlewares.
+    #
+    # Default implementations for `enter`, `leave` and `error`
+    # return the context object `ctx` without manipulating it.
+    # The subclasses can override all or some of these methods.
```

Comment:
I prefer middleware that only implements what it needs. It's unnecessary to call the pass though implementations and it makes is possible to look at middleware and dynamically tell what type of middleware it is (might be useful for debugging, logging, validating something, etc.).